### PR TITLE
Plugins feature and rework database

### PR DESCRIPTION
**Pull request by nyagami**
# **Bump React Native to 0.69 and Expo to 46**

## Why upgrade?
- Because old Hermes doenst support async functions
- Expo 46 is best compatible with RN 0.69

## Changes follow Reace Native Upgrade Helper
- ~~.editorconfig~~ 🚮
- ~~.gitattributes~~ 🚮
- ~~android/app/src/main/java/com/rajarsheechatterjee/LNReader/generated/BasePackageList.java~~ 🚮
- android/app/src/main/res/drawable/rn_edit_text_material.xml ➕
- the other modified files in android
## Changes follow Expo migration (from Expo 43)
- ~~package react-native-unimodules~~ and related configurations in android 🚮
- the other modified files in android
## Solving conflicting
 (that mean from now we just run `npm install`, no need `--leacy$*#$^#^%$#%`)
- ~~@react-native-community/masked-view~~🚮
- ~~lottie-react-native~~ 🚮
- ~~prop-types~~ 🚮
- ~~react-dom~~ 🚮
- ~~redux~~ 🚮
- ~~react-native-markdown-display~~ 🚮 (this is only used by Update dialog, and ... json file doest have any tag to markdown)
- the other modified packages (some follow RN, Expo, some I modify manually)
# App status
1. Search novels ✅
2. Parse novel and chapters ✅
3. Parsechapter ✅
4. Download chapters ✅
5. Deleting chapters ✅
6. History ✅
7. Edit novel info ✅
8. Add novel to Libary ✅ (click heart icon and long press)
9. Add categories for novels ✅
10. Filter, sort, ✅ (maybe)
11. Migrate novels ✅
12. Update novels, library ✅
13. Backup and Restore ✅
14. Restore only error novels ✅ 
 if you restored 98/100 novels successfully, you just need to retore the two left, not whole 100 novels
 15. Browse plugins (sources) ✅
		After updating plugins, github take a few minutes to update raw version of plugin
15. Statistics ✅
16. Settings ✅
17. Webview ✅
18. AnimeList ✅
19. Global search ✅
20. Hermes engine ✅
21. customJS, CSS ✅
22. Fullscreen mode ✅
23. Auto detect defalut language ✅
24. VolumeButton scroll amount ❌
25. restore from old database ❌
26. Bypass with cookies ✅
27. .... The other I havent checked.
# Notes
- **Hermes** support async functions `async function f(){}`, not arrow async functions ~`const f = () => {}`~
- Recommended environments: JDK 15, Nodejs 16, Android API 31
- What de hel is this warning? 
image
- Cannot auto scroll to previous position if chapter is long (solved ✅)

### CODE REVIEW

File: App.js [hunk 1]
```diff
@@ -13,12 +13,13 @@ import { Provider } from 'react-redux';
...
 
-import { createDatabase } from '@database/db';
+import { createTables } from '@database/db';
 import { persistor, store } from '@redux/store';
 import AppErrorBoundary from '@components/AppErrorBoundary/AppErrorBoundary';
+import { collectPlugins } from '@plugins/pluginManager';
 
 import Main from './src/navigators/Main';
-
+createTables();
```
[the remark this answers is not in the document]
Comment (nyagami) [reply App.js:22]: It was my mistake. I thought database was created before app ready to run.

File: src/components/NovelCover.js [hunk 2]
```diff
@@ -47,10 +46,9 @@ const NovelCover = ({
     [numColumns],
   );
 
-  const selectNovel = () => onLongPress && onLongPress(item.novelId);
-
-  const uri = item.novelCover;
+  const selectNovel = () => onLongPress(item);
```
[the remark this answers is not in the document]
Comment (nyagami) [reply src/components/NovelCover.js:49]: When is onLongPress undefined?

File: src/plugins/helpers/parseDate.js [hunk 3]
```diff
@@ -21,7 +21,7 @@ export const parseMadaraDate = date => {
       releaseDate.setMonth(releaseDate.getMonth() - timeAgo);
     }
 
-    releaseDate = dayjs(releaseDate).format('LL');
+    releaseDate = dayjs(releaseDate).format('DD/MM/YYYY');
```
[the remark this answers is not in the document]
Comment (nyagami) [reply src/plugins/helpers/parseDate.js:24]: The source I tested on didn't work with this format. I will test it again

File: src/components/NovelCover.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (rajarsheechatterjee) [reply src/components/NovelCover.js:49]: I think it used to be when browse screen didn't have long press to add to library. Not required anymore I guess.

**Comment by nyagami** [comment]
[488f450](https://github.com/LNReader/lnreader/pull/579/commits/488f450205b3cb117d5fcba56c58cc3c555bdb68)
This should be removed. Because novel and chapters must be fetched together or nothing.
Filter and **Transperant model** may cause disappearing novels and chapters problem.

**Comment by nyagami** [comment]
# **Bump React Native to 0.69 and Expo to 46**

## Why upgrade?
- Because old Hermes doenst support async functions
- Expo 46 is best compatible with RN 0.69

## Changes follow Reace Native Upgrade Helper
- ~~.editorconfig~~ 🚮
- ~~.gitattributes~~ 🚮
- ~~android/app/src/main/java/com/rajarsheechatterjee/LNReader/generated/BasePackageList.java~~ 🚮
- android/app/src/main/res/drawable/rn_edit_text_material.xml ➕
- the other modified files in android
## Changes follow Expo migration (from Expo 43)
- ~~package react-native-unimodules~~ and related configurations in android 🚮
- the other modified files in android
## Solving conflicting
 (that mean from now we just run `npm install`, no need `--leacy$*#$^#^%$#%`)
- ~~@react-native-community/masked-view~~🚮
- ~~lottie-react-native~~ 🚮
- ~~prop-types~~ 🚮
- ~~react-dom~~ 🚮
- ~~redux~~ 🚮
- ~~react-native-markdown-display~~ 🚮 (this is only used by Update dialog, and ... json file doest have any tag to markdown)
- the other modified packages (some follow RN, Expo, some I modify manually)
# App status
1. Search novels ✅
2. Parse novel and chapters ✅
3. Parsechapter ✅
4. Download chapters ✅
5. Deleting chapters ✅
6. History ✅
7. Edit novel info ✅
8. Add novel to Libary ✅ (click heart icon and long press)
9. Add categories for novels ✅
10. Filter, sort, ✅ (maybe)
11. Migrate novels ❌ (wrong function, ig)
12. Update novels, library ❔(dont know how to test)
13. Backup and Restore ✅
14. Restore only error novels ✅ 
 if you restored 98/100 novels successfully, you just need to retore the two left, not whole 100 novels
 15. Browse plugins (sources) ❔
		Sometimes, I cant fetch newtest version of plugin (dont know why)
		Display problems (related to redux, could be fixed)
15. Statistics ✅
16. Settings ✅
17. Webview ✅
18. AnimeList ✅
19. Global search ❔(havent tested)
20. .... The other I havent checked.
# Notes
- **Hermes** support async functions `async function f(){}`, not arrow async functions ~`const f = () => {}`~
- Maybe `Regex` cant be used in plugin scripts.
- What de hel this warning? 
image

**Comment by nyagami** [comment]
fetchFiles are different because RN supports `FileReader` but not `BufferReader`, the oposite, Nodejs supports `BufferReader` but not `FileReader` 😂

**Comment by nyagami** [comment]
Before testing, remember to clear all app's data. 😎

**Pull request merged**